applications/IgaApplication/custom_conditions/support_penalty_condition.cpp

```diff
@@ -0,0 +1,176 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Ricky Aristio
+//                   Tobias Teschemacher
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "custom_conditions/support_penalty_condition.h"
+
+namespace Kratos
+{
+    void SupportPenaltyCondition::CalculateAll(
+        MatrixType& rLeftHandSideMatrix,
+        VectorType& rRightHandSideVector,
+        const ProcessInfo& rCurrentProcessInfo,
+        const bool CalculateStiffnessMatrixFlag,
+        const bool CalculateResidualVectorFlag
+    )
+    {
+        KRATOS_TRY
+        const double penalty = GetProperties()[PENALTY_FACTOR];
+
+        const auto& r_geometry = GetGeometry();
+        const SizeType number_of_nodes = r_geometry.size();
+        const SizeType mat_size = r_geometry.WorkingSpaceDimension() * number_of_nodes;
+
+        // Integration
+        const GeometryType::IntegrationPointsArrayType& integration_points = r_geometry.IntegrationPoints();
+
+        // initial determinant of jacobian 
+        Vector determinat_jacobian_vector_initial(integration_points.size());
+        DeterminantOfJacobianInitial(r_geometry, determinat_jacobian_vector_initial);
+
+        for (IndexType point_number = 0; point_number < integration_points.size(); point_number++)
+        {
+            const Matrix& N = r_geometry.ShapeFunctionsValues();
+
+            //FOR DISPLACEMENTS
+            Matrix H = ZeroMatrix(3, mat_size);
+            for (IndexType i = 0; i < number_of_nodes; i++)
+            {
+                IndexType index = 3 * i;
+                H(0, index)     = N(point_number, i);
+                H(1, index + 1) = N(point_number, i);
+                H(2, index + 2) = N(point_number, i);
+            }
+
+            // Differential area
+            const double integration_weight = integration_points[point_number].Weight();
+            const double determinat_jacobian = determinat_jacobian_vector_initial[point_number];
+
+            // Assembly
+            if (CalculateStiffnessMatrixFlag) {
+                noalias(rLeftHandSideMatrix) += prod(trans(H), H)
+                    * penalty * integration_weight * determinat_jacobian;
+            }
+            if (CalculateResidualVectorFlag) {
+
+                const array_1d<double, 3>& displacement = Has(DISPLACEMENT)
+                    ? this->GetValue(DISPLACEMENT)
+                    : ZeroVector(3);
+
+                Vector u(mat_size);

```

Comment:
pre-initialize the u vector, if possible

Reply:
I think it is more optimized in that way, as initializing is also a value assignment which should take some processing time.